Scripts/Core/EditorVR.Menus.cs

```diff
@@ -197,119 +109,282 @@ internal void UpdateMenuVisibilities()
-				for (int i = 0; i < m_ActiveDeviceData.Count; i++)
+				var directSelection = evr.GetNestedModule<DirectSelection>();
+
+				foreach (var deviceData in m_ActiveDeviceData)
 				{
-					var deviceData = m_ActiveDeviceData[i];
 					var alternateMenu = deviceData.alternateMenu;
 					var mainMenu = deviceData.mainMenu;
 					var customMenu = deviceData.customMenu;
 					var menuHideFlags = deviceData.menuHideFlags;
 
-					// Move alternate menu to another device if it conflicts with main or custom menu
-					if (alternateMenu != null && (menuHideFlags[mainMenu] == 0 || (customMenu != null && menuHideFlags[customMenu] == 0)) && menuHideFlags[alternateMenu] == 0)
+					var mainMenuVisible = mainMenu != null && (menuHideFlags[mainMenu] & MenuHideFlags.Hidden) == 0;
+					var customMenuVisible = customMenu != null && (menuHideFlags[customMenu] & MenuHideFlags.Hidden) == 0 && (menuHideFlags[customMenu] & MenuHideFlags.OtherMenu) == 0;
+					var alternateMenuVisible = alternateMenu != null && (menuHideFlags[alternateMenu] & MenuHideFlags.Hidden) == 0;
+
+					if (alternateMenuVisible && (mainMenuVisible || customMenuVisible))
 					{
-						for (int j = 0; j < m_ActiveDeviceData.Count; j++)
+						foreach (var otherDeviceData in m_ActiveDeviceData)
 						{
-							var otherDeviceData = m_ActiveDeviceData[j];
 							if (otherDeviceData == deviceData)
 								continue;
 
-							var otherCustomMenu = otherDeviceData.customMenu;
-							var otherHideFlags = otherDeviceData.menuHideFlags;
-							otherHideFlags[otherDeviceData.alternateMenu] &= ~MenuHideFlags.Hidden;
-							otherHideFlags[otherDeviceData.mainMenu] |= MenuHideFlags.Hidden;
+							SetAlternateMenuVisibility(otherDeviceData.rayOrigin, true);
+							break;
+						}
+					}
 
-							if (otherCustomMenu != null)
-								otherHideFlags[otherCustomMenu] |= MenuHideFlags.Hidden;
+					if (customMenuVisible && (mainMenuVisible || alternateMenuVisible))
+						menuHideFlags[customMenu] |= MenuHideFlags.OtherMenu;
+
+					Array.Clear(m_WorkspaceOverlaps, 0, m_WorkspaceOverlaps.Length);
+					var hoveringWorkspace = false;
+					var menuTransform = mainMenu.menuContent.transform;
+					var menuBounds = mainMenu.localBounds;
+					var menuRotation = menuTransform.rotation;
+					var viewerScale = Viewer.GetViewerScale();
+					var center = menuTransform.position + menuRotation * menuBounds.center * viewerScale;
+					if (Physics.OverlapBoxNonAlloc(center, menuBounds.extents * viewerScale, m_WorkspaceOverlaps, menuRotation) > 0)
+					{
+						foreach (var overlap in m_WorkspaceOverlaps)
+						{
+							if (overlap)
+							{
+								m_WorkspaceComponents.Clear();
+								overlap.GetComponents(m_WorkspaceComponents);
+								if (m_WorkspaceComponents.Count > 0)
+									hoveringWorkspace = true;
+							}
 						}
+					}
 
-						menuHideFlags[alternateMenu] |= MenuHideFlags.Hidden;
+					var menus = menuHideFlags.Keys.ToList();
```

Comment:
The fix here was to combine hideflags and hide/show time into a MenuHideData class, so now we can set fields on those classes and we're not modifying the dictionary.

It looks like that last snippet in my previous comment is the best/only way to modify dictionary values in a loop without allocating.